README.md
![whetstone banner](https://cdn.modrinth.com/data/M97mSTbc/images/41334af848e5845f490b391c1765e8285021f1ad.png)

### <center>Whetstones</center>
---

**Whetstones** allows you to easily repair tools in hand. You can use whetstones by holding a whetstone in one hand and a tool in the other hand, then right-clicking it.

CurseForge: https://www.curseforge.com/minecraft/mc-mods/whetstones  
Modrinth: https://modrinth.com/mod/whetstones

---

### <img width='24' src='https://imgur.com/IV4nmmQ.png' alt='Wooden whetstone'> Wooden Whetstone
- Can repair wood tier tools and lower
- Can repair 59 durability

### <img width='24' src='https://imgur.com/TQZZsgT.png' alt='Stone whetstone'> Stone Whetstone
- Can repair stone tier tools and lower
- Can repair 131 durability

### <img width='24' src='https://imgur.com/tTEnaHh.png' alt='Iron whetstone'> Iron Whetstone
- Can repair iron tier tools and lower
- Can repair 250 durability

### <img width='24' src='https://imgur.com/E67CMpP.png' alt='Golden whetstone'> Golden Whetstone
- Can repair gold tier tools and lower
- Can repair 32 durability

### <img width='24' src='https://imgur.com/hB1vuDw.png' alt='Diamond whetstone'> Diamond Whetstone
- Can repair diamond tier tools and lower
- Can repair 1561 durability

### <img width='24' src='https://imgur.com/FpjZ6lt.png' alt='Netherite whetstone'> Netherite Whetstone
- Can repair any tier tools
- Can repair 2031 durability

---

### How to use

1. Equip the tool you want to repair in the off-hand.
2. Hold the whetstone in your main hand.
3. Right-click to repair the tool

![Repairing diamond pickaxe](https://imgur.com/mzD9zjy.gif)

---

### Config options

**'onlyTools': true/false**   
If true, only tools can be repaired, so no other items like armor.

**'onlyLowerTiers': true/false**   
If true, only tools with a tier lower than or equal to the whetstone can be repaired. Netherite can repair all tiers. Only takes effect if 'onlyTools' is enabled.

**Durability per whetstone type: a value between 1 and 10000**   
How much durability should the whetstone type have?

---

### FAQ

**Can I use your mod in my modpack?**   
Yes, feel free to use my mod in your modpack

---

![whetstone banner](https://cdn.modrinth.com/data/M97mSTbc/images/41334af848e5845f490b391c1765e8285021f1ad.png)


changelog.md
### Whetstones 1.0.1
- Fixed repair for wooden and stone whetstones

### Whetstones 1.0.0
- Initial version


src/main/java/com/bakabreak/whetstones/Whetstones.java
package com.bakabreak.whetstones;

import com.bakabreak.whetstones.generators.WhetstoneRecipeGenerator;
import com.bakabreak.whetstones.generators.WhetstonesLanguageGenerator;
import com.bakabreak.whetstones.generators.WhetstonesModelGenerator;
import com.supermartijn642.core.item.BaseItem;
import com.supermartijn642.core.item.CreativeItemGroup;
import com.supermartijn642.core.item.ItemProperties;
import com.supermartijn642.core.registry.GeneratorRegistrationHandler;
import com.supermartijn642.core.registry.RegistrationHandler;
import com.supermartijn642.core.registry.RegistryEntryAcceptor;
import net.minecraftforge.fml.common.Mod;

@Mod("whetstones")
public class Whetstones {

    @RegistryEntryAcceptor(namespace = "whetstones", identifier = "diamond_whetstone", registry = RegistryEntryAcceptor.Registry.ITEMS)
    public static BaseItem diamondWhetstone;

    public static final CreativeItemGroup GROUP = CreativeItemGroup.create("whetstones", () -> diamondWhetstone);

    public Whetstones() {
        RegistrationHandler handler = RegistrationHandler.get("whetstones");
        for (WhetstoneTiers tier : WhetstoneTiers.values())
            handler.registerItemCallback(tier::createItem);
        GeneratorRegistrationHandler generatorHandler = GeneratorRegistrationHandler.get("whetstones");
        generatorHandler.addGenerator(WhetstonesLanguageGenerator::new);
        generatorHandler.addGenerator(WhetstonesModelGenerator::new);
        generatorHandler.addGenerator(WhetstoneRecipeGenerator::new);
    }
}


src/main/java/com/bakabreak/whetstones/WhetstoneItem.java
package com.bakabreak.whetstones;

import com.supermartijn642.core.TextComponents;
import com.supermartijn642.core.item.BaseItem;
import com.supermartijn642.core.item.ItemProperties;
import net.minecraft.ChatFormatting;
import net.minecraft.network.chat.Component;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.TieredItem;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.Level;
import net.minecraftforge.common.TierSortingRegistry;
import org.jetbrains.annotations.Nullable;

import java.util.function.Consumer;

public class WhetstoneItem extends BaseItem {
    private final WhetstoneTiers tier;

    public WhetstoneItem(ItemProperties properties, WhetstoneTiers tier) {
        super(properties);
        this.tier = tier;
    }

    @Override
    protected void appendItemInformation(ItemStack stack, @Nullable BlockGetter level, Consumer<Component> info, boolean advanced) {
        info.accept(this.tier.getDescriptionTranslation());
        Component durability = TextComponents.number(stack.getMaxDamage() - stack.getDamageValue()).color(ChatFormatting.GOLD).get();
        if (stack.getDamageValue() > 0)
            info.accept(TextComponents.translation("whetstones.item.description.durability.used", durability).color(ChatFormatting.GRAY).get());
        else
            info.accept(TextComponents.translation("whetstones.item.description.durability", durability).color(ChatFormatting.GRAY).get());
    }

    @Override
    public ItemUseResult interact(ItemStack stack, Player player, InteractionHand hand, Level level) {
        ItemStack repairItem = player.getItemInHand(hand == InteractionHand.MAIN_HAND ? InteractionHand.OFF_HAND : InteractionHand.MAIN_HAND);
        if (repairItem.getItem().canBeDepleted() && repairItem.isDamaged() && (!WhetstonesConfig.onlyTools.get() || (repairItem.getItem() instanceof TieredItem && (!WhetstonesConfig.onlyLowerTiers.get() || (this.tier.tier == null || this.tier.tier == ((TieredItem) repairItem.getItem()).getTier() || TierSortingRegistry.getTiersLowerThan(this.tier.tier).contains(((TieredItem) repairItem.getItem()).getTier())))))) {

            int repairAmount = (int) Math.max(1, Math.min(repairItem.getDamageValue(), stack.getMaxDamage() / 75d));
            repairItem.setDamageValue(repairItem.getDamageValue() - repairAmount);
            if (!player.isCreative())
                stack.hurtAndBreak(repairAmount, player, p -> p.broadcastBreakEvent(hand));
            return ItemUseResult.consume(stack);
        }
        return ItemUseResult.fail(stack);
    }
}


src/main/java/com/bakabreak/whetstones/WhetstoneTiers.java
package com.bakabreak.whetstones;

import com.supermartijn642.core.TextComponents;
import com.supermartijn642.core.item.ItemProperties;
import com.supermartijn642.core.registry.RegistrationHandler;
import net.minecraft.ChatFormatting;
import net.minecraft.network.chat.Component;
import net.minecraft.tags.ItemTags;
import net.minecraft.tags.TagKey;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.Tier;
import net.minecraft.world.item.Tiers;
import net.minecraftforge.common.Tags;

import java.awt.*;
import java.util.Locale;
import java.util.function.Supplier;

public enum WhetstoneTiers {
    WOODEN("Wooden Whetstone", "Restores durability for wooden tools and below", WhetstonesConfig.woodenWhetstoneDurability, Tags.Items.RODS_WOODEN, Tiers.WOOD),
    STONE("Stone Whetstone", "Restores durability for stone tools and below", WhetstonesConfig.stoneWhetstoneDurability, ItemTags.STONE_CRAFTING_MATERIALS, Tiers.STONE),
    IRON("Iron Whetstone", "Restores durability for iron tools and below", WhetstonesConfig.ironWhetstoneDurability, Tags.Items.INGOTS_IRON, Tiers.IRON),
    GOLDEN("Golden Whetstone", "Restores durability for golden tools and below", WhetstonesConfig.goldenWhetstoneDurability, Tags.Items.INGOTS_GOLD, Tiers.GOLD),
    DIAMOND("Diamond Whetstone", "Restores durability for diamond tools and below", WhetstonesConfig.diamondWhetstoneDurability, Tags.Items.GEMS_DIAMOND, Tiers.DIAMOND),
    NETHERITE("Netherite Whetstone", "Restores durability for any tools", WhetstonesConfig.netheriteWhetstoneDurability, Tags.Items.INGOTS_NETHERITE, null);

    public final String displayName, description;
    private final Supplier<Integer> durability;
    public final TagKey<Item> craftingMaterial;
    private WhetstoneItem item;
    public final Tier tier;

    WhetstoneTiers(String displayName, String description, Supplier<Integer> durability, TagKey<Item> craftingMaterial, Tier tier) {
        this.displayName = displayName;
        this.description = description;
        this.durability = durability;
        this.craftingMaterial = craftingMaterial;
        this.tier = tier;
    }

    public String getIdentifier() {
        return this.name().toLowerCase(Locale.ROOT) + "_whetstone";
    }

    public int getDurability() {
        return this.durability.get();
    }

    public void createItem(RegistrationHandler.Helper<Item> helper) {
        this.item = new WhetstoneItem(ItemProperties.create().group(Whetstones.GROUP).durability(this.getDurability()), this);
        helper.register(this.getIdentifier(), this.item);
    }

    public Component getDescriptionTranslation(){
        return TextComponents.translation("whetstones." + this.getIdentifier() + ".description").color(ChatFormatting.DARK_GRAY).get();
    }

    public WhetstoneItem getItem() {
        if (this.item == null)
            throw new RuntimeException("Item has not yet been created!!");
        return item;
    }
}


src/main/java/com/bakabreak/whetstones/generators/WhetstoneRecipeGenerator.java
package com.bakabreak.whetstones.generators;

import com.bakabreak.whetstones.WhetstoneTiers;
import com.supermartijn642.core.generator.RecipeGenerator;
import com.supermartijn642.core.generator.ResourceCache;
import net.minecraft.tags.ItemTags;
import net.minecraft.world.item.Items;

public class WhetstoneRecipeGenerator extends RecipeGenerator {
    public WhetstoneRecipeGenerator(ResourceCache cache) {
        super("whetstones", cache);
    }

    @Override
    public void generate() {
        for (WhetstoneTiers tier : WhetstoneTiers.values()) {
            this.shaped(tier.getItem())
                    .pattern("#")
                    .pattern("$")
                    .input('#', tier.craftingMaterial)
                    .input('$', ItemTags.PLANKS)
                    .unlockedBy(tier.craftingMaterial);
        }
    }
}


src/main/java/com/bakabreak/whetstones/generators/WhetstonesLanguageGenerator.java
package com.bakabreak.whetstones.generators;

import com.bakabreak.whetstones.WhetstoneTiers;
import com.bakabreak.whetstones.Whetstones;
import com.supermartijn642.core.generator.LanguageGenerator;
import com.supermartijn642.core.generator.ResourceCache;

public class WhetstonesLanguageGenerator extends LanguageGenerator {
    public WhetstonesLanguageGenerator(ResourceCache cache) {
        super("whetstones", cache, "en_us");
    }

    @Override
    public void generate() {
        for (WhetstoneTiers tier : WhetstoneTiers.values()) {
            this.item(tier.getItem(), tier.displayName);
            this.translation("whetstones." + tier.getIdentifier() + ".description", tier.description);
        }
        this.translation("whetstones.item.description.durability", "Durability: %s");
        this.translation("whetstones.item.description.durability.used", "Remaining durability: %s");
        this.itemGroup(Whetstones.GROUP, "Whetstones");
    }
}


src/main/java/com/bakabreak/whetstones/WhetstonesConfig.java
package com.bakabreak.whetstones;

import com.supermartijn642.configlib.ConfigBuilder;
import com.supermartijn642.configlib.api.ConfigBuilders;
import com.supermartijn642.configlib.api.IConfigBuilder;

import java.util.function.Supplier;

public class WhetstonesConfig {
    public static final Supplier<Integer> woodenWhetstoneDurability;
    public static final Supplier<Integer> stoneWhetstoneDurability;
    public static final Supplier<Integer> ironWhetstoneDurability;
    public static final Supplier<Integer> goldenWhetstoneDurability;
    public static final Supplier<Integer> diamondWhetstoneDurability;
    public static final Supplier<Integer> netheriteWhetstoneDurability;
    public static final Supplier<Boolean> onlyTools;
    public static final Supplier<Boolean> onlyLowerTiers;

    static {
        IConfigBuilder builder = ConfigBuilders.newTomlConfig("whetstones", null, false);
        builder.push("General");
        onlyTools = builder.comment("If true, only tools can be repaired, so no other items like armor.").define("onlyTools", true);
        onlyLowerTiers = builder.comment("If true, only tools with a tier lower than or equal to the whetstone can be repaired. Netherite can repair all tiers. Only takes effect if 'onlyTools' is enabled.").define("onlyLowerTiers", true);
        builder.pop();
        builder.push("Durability");
        woodenWhetstoneDurability = builder.comment("How much durability should the wooden whetstone restore?").gameRestart().define("woodenWhetstoneDurability", 59, 1, 10000);
        stoneWhetstoneDurability = builder.comment("How much durability should the stone whetstone restore?").gameRestart().define("stoneWhetstoneDurability", 131, 1, 10000);
        ironWhetstoneDurability = builder.comment("How much durability should the iron whetstone restore?").gameRestart().define("ironWhetstoneDurability", 250, 1, 10000);
        goldenWhetstoneDurability = builder.comment("How much durability should the golden whetstone restore?").gameRestart().define("goldenWhetstoneDurability", 32, 1, 10000);
        diamondWhetstoneDurability = builder.comment("How much durability should the diamond whetstone restore?").gameRestart().define("diamondWhetstoneDurability", 1561, 1, 10000);
        netheriteWhetstoneDurability = builder.comment("How much durability should the netherite whetstone restore?").gameRestart().define("netheriteWhetstoneDurability", 2031, 1, 10000);
        builder.pop();
        builder.build();
    }
}
